### DISCUSSION TITLE
Secure storage of sessions in localStorage using encryption

### DISCUSSION THREAD

**Discussion by koddr** [supabase/supabase#3731] [Feature Requests]
Hi,

First, thank you for Supabase! This is a great, helpful service that I've been looking for years.

I noticed that all data of the current user session saved in `localStorage` when using the Supabase JavaScript client. This is not secure for many reasons (_I think you all know them already, I won't repeat them_).

> I'm not saying you have to use the libraries that I give as an example below! I'm advocating the very principle of introducing security into public areas. Our safety is in our hands.

For example, [`softvar/secure-ls`](https://github.com/softvar/secure-ls) library (the first result in a Google search for `secure localStorage`), which uses [these](https://github.com/softvar/secure-ls#libraries-used) libraries to solve a problem: 

- https://code.google.com/archive/p/crypto-js/
- https://github.com/pieroxy/lz-string

In our case, Supabase JS client could do the following (under the hood): 

```js
// Init secure-ls (for example):
let ls = new SecureLS();

// Setting data of the current session:
ls.set('supabase.auth.token', { currentSession: { ... }  })

// For getting data of the current session:
ls.get('supabase.auth.token');
```

But for users, the value of the `supabase.auth.token` in the `localStorage` is still encrypted.

Another solution, but on the server side. 

Save important (secret) information about session in `HttpOnly` cookie, which will be checked by Supabase backend at each request. And on the browser side, in the `localStorage` store a random key that will tell the Supabase JavaScript client that the user has been authorized (for making requests).

Again, for example, the Fiber Go web framework ([`gofiber/fiber`](https://github.com/gofiber/fiber)) has a built-in middleware [`encryptcookie`](https://github.com/gofiber/fiber/tree/master/middleware/encryptcookie) to do this:

```go
// Middleware config:
app.Use(encryptcookie.New(encryptcookie.Config{
    Key: encryptcookie.GenerateKey(),
}))

// Create the encrypted cookie with current session:
app.Post("/login", func(c *fiber.Ctx) error {
    c.Cookie(&fiber.Cookie{
        Name:  "supabase.auth.token",
        Value: supabase.CreateCurrentSession{ ... },
    })
    return nil
})

// Reading the encrypted cookie with current session:
app.Get("/dashboard", func(c *fiber.Ctx) error {
    return c.SendString(c.Cookies("supabase.auth.token"))
})
```

Once again, we have a protected value that cannot be decrypted in any way without the key that the backend generates.

**Comment by fergusmeiklejohn** [comment]
Many possible solutions.. It's a problem for sure, especially for apps with sensitive user data.
We're using Remix as our React framework https://remix.run, where the the default is to do all our auth on the server. eg: https://github.com/sergiodxa/remix-auth
ATM the supabase auth solution make it *seem* like doing auth server side would be difficult.

**Reply by koddr** [reply]
Hi,

If I understand you correctly, are you suggesting trying to bring `localStrorage` encryption to the level of my application? That is, do an additional wrapper over `supabase.auth.session()` or something like that?

**Reply by fergusmeiklejohn** [reply]
I meant like we use secure cookies and we don't use local storage at all, (or at least that's a possibility for those apps which need it)

**Reply by koddr** [reply]
Oh, I'm sorry, yeah now I've re-read it again, and I get it!

Yes, it would be great to have an HttpOnly cookie with information to authorize requests to the Supabase server. 

I hope the developers will read this thread 😉

**Comment by acatataca** [comment]
Are there any updates on this topic since Nov 21? I'm looking for best practices for securing sessions with supabase and remix.

Thanks,
Sam

**Comment by hf** [comment]
Hey, I'm from the Supabase Auth team. I'd like explain a bit on the topics in the discussion here. (We're going to be releasing some official documentation on Security soon which will explain other topics as well.)

First, on security of cookies and session information. GoTrue (the open-source server behind Supabase Auth) issues two pieces of information that identify a user:

1. An access token (in the form of a JWT)
2. A refresh token (an opaque string)

Both of these are set as cookies on `<project>.supabase.co`. Furthermore, both are given to your frontend JavaScript in a callback within the URL fragment.

All applications using Supabase Auth need access to both of these values at all times, so that they can keep a user's session alive and retrieve important information about the user's profile.

This information must be stored somewhere in the browser for the client libraries to work. Our security model (which is typical for most browser-based applications) does not account for a breached device or browser. This is because there is virtually nothing that can be done on a technical level to prevent security issues if the end-user's device is compromised. Adding additional levels of "encryption" only adds another layer of obfuscation, which in and of itself does not provide additional security.

There are application-level measures that can be done to reduce security risks of compromised devices. For example, asking for re-authentication when the user performs sensitive actions, or controlling the total duration of a session by instituting a maximum session duration and inactivity-based session duration. Finally, adding multi-factor authentication capabilities to the application is always a good idea.

Supabase Auth unfortunately does not currently have features that support some of these application-level measures, but we are working very hard to make them happen. Do watch the [GoTrue](https://github.com/supabase/gotrue) repository to follow with the latest updates.

**Reply by fergusmeiklejohn** [reply]
Thanks that's very helpful!
This question comes up a lot in the Remix forum because the way Remix works, it leans towards the server as the single source of truth. So authenticating on the server and storing auth in a session cookie is the happy path with Remix. This is a challenge with Supabase which leans towards a client state model. So from the POV of Remix and other similar apps, it would be best to auth with Supabase on the server and pass the access token to the client to write it to local storage so Supabase's client SDK will work.
It would be very helpful to document what we need to do to make this authentication model work. What exactly do we need to write to localhost?
